src/VisualStudio/LiveShare/Impl/RunCodeActionsHandler.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.LanguageServer.CustomProtocol;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.VisualStudio.LiveShare.LanguageServices;
+using Newtonsoft.Json.Linq;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.VisualStudio.LanguageServices.LiveShare
+{
+    /// <summary>
+    /// Run code actions handler.  Called when lightbulb invoked.
+    /// Code actions must be applied from the UI thread in VS.
+    /// </summary>
+    internal class RunCodeActionsHandler : CodeActionsHandlerBase, ILspRequestHandler<LSP.ExecuteCommandParams, object, Solution>
+    {
+        private IThreadingContext _threadingContext;
+
+        [ImportingConstructor]
+        public RunCodeActionsHandler(ICodeFixService codeFixService, ICodeRefactoringService codeRefactoringService, IThreadingContext threadingContext)
+            : base(codeFixService, codeRefactoringService)
+        {
+            _threadingContext = threadingContext;
+        }
+
+        public async Task<object> HandleAsync(LSP.ExecuteCommandParams request, RequestContext<Solution> requestContext, CancellationToken cancellationToken)
+        {
+            // Unwrap the command to get the RunCodeActions command.
+            if (request.Command.StartsWith(CodeActionsHandlerShim.RemoteCommandNamePrefix))
+            {
+                var command = ((JObject)request.Arguments[0]).ToObject<LSP.Command>();
+                request.Command = command.CommandIdentifier;
+                request.Arguments = command.Arguments;
+            }
+
+            if (request.Command == RunCodeActionCommandName)
+            {
+                var runRequest = ((JToken)request.Arguments.Single()).ToObject<RunCodeActionParams>();
+                var codeActions = await GetCodeActionsAsync(requestContext.Context,
+                    runRequest.CodeActionParams.TextDocument.Uri,
+                    runRequest.CodeActionParams.Range,
+                    cancellationToken).ConfigureAwait(false);
+
+                var actionToRun = codeActions?.FirstOrDefault(a => a.Title == runRequest.Title);
+
+                if (actionToRun != null)
+                {
+                    // TODO - This UI thread dependency must be removed.
```

Comment:
Is this actually a TODO at this point? Invoking code fixes is a guarantee, so it's not really something _easily_ removable.